File: packages/nerva-py/nerva/memory/cold.py
"""Cold memory — long-term knowledge with search capability.

In-memory implementation of the ``ColdTier`` protocol from ``tiered.py``.
Uses simple word-overlap scoring instead of vector embeddings. Suitable
for testing; swap with a vector DB implementation for production.
"""

from __future__ import annotations

from collections import defaultdict

DEFAULT_MAX_RESULTS = 10
"""Maximum search results returned per query."""

MIN_RELEVANCE_SCORE = 0.1
"""Minimum word-overlap score for an entry to appear in search results."""


class InMemoryColdMemory:
    """In-memory cold tier with basic keyword search.

    No actual vector embeddings — uses simple word-overlap scoring
    (|query intersection entry| / |query words|) to rank stored entries.
    For production use, swap with a vector DB implementation.

    Args:
        max_results: Maximum number of search results to return.
    """

    def __init__(self, max_results: int = DEFAULT_MAX_RESULTS) -> None:
        self._max_results = max_results
        self._entries: dict[str, list[str]] = defaultdict(list)

    async def search(self, query: str, scope: str) -> list[str]:
        """Search stored entries by word overlap with the query.

        Scores each entry as ``|intersection| / |query_words|``.
        Returns the top-k entries above ``MIN_RELEVANCE_SCORE``,
        sorted by score descending.

        Args:
            query: Search query text.
            scope: Memory scope string for filtering.

        Returns:
            List of matching knowledge strings, most relevant first.
            At most ``max_results`` entries are returned.
        """
        query_words = _to_word_set(query)
        if not query_words:
            return []

        entries = self._entries.get(scope, [])
        scored = _score_entries(query_words, entries)
        return _select_top_results(scored, self._max_results)

    async def store(self, content: str, scope: str) -> None:
        """Store a knowledge entry in the cold tier.

        Duplicate entries (exact match) within the same scope are skipped.

        Args:
            content: Knowledge text to store.
            scope: Memory scope string for isolation.
        """
        if not content or not content.strip():
            return

        entries = self._entries[scope]
        if content not in entries:
            entries.append(content)

    async def clear(self, scope: str) -> None:
        """Remove all entries for a scope.

        Args:
            scope: Memory scope to clear.
        """
        self._entries.pop(scope, None)


# ---------------------------------------------------------------------------
# Scoring helpers
# ---------------------------------------------------------------------------


def _score_entries(
    query_words: set[str], entries: list[str],
) -> list[tuple[float, str]]:
    """Score entries by word overlap with pre-computed query words.

    Args:
        query_words: Lowercase word set from the search query.
        entries: Candidate knowledge strings.

    Returns:
        List of (score, entry) tuples above the relevance threshold.
    """
    scored: list[tuple[float, str]] = []
    for entry in entries:
        score = _word_overlap_score(query_words, entry)
        if score >= MIN_RELEVANCE_SCORE:
            scored.append((score, entry))
    return scored


def _select_top_results(
    scored: list[tuple[float, str]], max_results: int,
) -> list[str]:
    """Sort scored entries descending and return top-k texts.

    Args:
        scored: List of (score, entry) tuples.
        max_results: Maximum entries to return.

    Returns:
        Entry texts sorted by relevance, capped at max_results.
    """
    scored.sort(key=lambda item: -item[0])
    return [entry for _, entry in scored[:max_results]]


def _word_overlap_score(query_words: set[str], text: str) -> float:
    """Compute word-overlap ratio between query words and text.

    Args:
        query_words: Pre-computed lowercase word set from the query.
        text: Text to score against the query.

    Returns:
        Ratio of overlapping words to total query words (0.0 to 1.0).
    """
    text_words = _to_word_set(text)
    if not text_words:
        return 0.0
    overlap = len(query_words & text_words)
    return overlap / len(query_words)


def _to_word_set(text: str) -> set[str]:
    """Split text into a lowercase word set.

    Args:
        text: Input text.

    Returns:
        Set of lowercase words (whitespace-split).
    """
    return set(text.lower().split())

File: packages/nerva-js/src/memory/cold.ts
/**
 * Cold memory — long-term knowledge with keyword search.
 *
 * In-memory implementation of the {@link ColdTier} protocol from tiered.ts.
 * Uses simple word-overlap scoring instead of vector embeddings. Suitable
 * for testing; swap with a vector DB implementation for production.
 *
 * @module memory/cold
 */

import type { ColdTier } from "./tiered.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Maximum search results returned per query. */
export const DEFAULT_MAX_RESULTS = 10;

/** Minimum word-overlap score for an entry to appear in search results. */
const MIN_RELEVANCE_SCORE = 0.1;

// ---------------------------------------------------------------------------
// InMemoryColdMemory
// ---------------------------------------------------------------------------

/** Configuration options for {@link InMemoryColdMemory}. */
export interface ColdMemoryOptions {
  /** Maximum number of search results to return. */
  readonly maxResults?: number;
}

/**
 * In-memory cold tier with basic keyword search.
 *
 * No actual vector embeddings — uses simple word-overlap scoring
 * (|query intersection entry| / |query words|) to rank stored entries.
 * For production use, swap with a vector DB implementation.
 */
export class InMemoryColdMemory implements ColdTier {
  private readonly _maxResults: number;
  private readonly _entries = new Map<string, string[]>();

  /**
   * @param options - Optional max results override.
   */
  constructor(options?: ColdMemoryOptions) {
    this._maxResults = options?.maxResults ?? DEFAULT_MAX_RESULTS;
  }

  /**
   * Search stored entries by word overlap with the query.
   *
   * Scores each entry as |intersection| / |query_words|.
   * Returns the top-k entries above the relevance threshold,
   * sorted by score descending.
   *
   * @param query - Search query text.
   * @param scope - Memory scope string for filtering.
   * @returns List of matching knowledge strings, most relevant first.
   */
  async search(query: string, scope: string): Promise<string[]> {
    const queryWords = toWordSet(query);
    if (queryWords.size === 0) {
      return [];
    }

    const entries = this._entries.get(scope) ?? [];
    const scored = scoreEntries(queryWords, entries);
    return selectTopResults(scored, this._maxResults);
  }

  /**
   * Store a knowledge entry in the cold tier.
   *
   * Duplicate entries (exact match) within the same scope are skipped.
   *
   * @param content - Knowledge text to store.
   * @param scope - Memory scope string for isolation.
   */
  async store(content: string, scope: string): Promise<void> {
    if (!content || !content.trim()) {
      return;
    }

    const entries = this.getOrCreate(scope);
    if (!entries.includes(content)) {
      entries.push(content);
    }
  }

  /**
   * Remove all entries for a scope.
   *
   * @param scope - Memory scope to clear.
   */
  async clear(scope: string): Promise<void> {
    this._entries.delete(scope);
  }

  // -- Private helpers ------------------------------------------------------

  /**
   * Get or create the entries array for a scope.
   *
   * @param scope - Memory scope key.
   * @returns The mutable entries array.
   */
  private getOrCreate(scope: string): string[] {
    const existing = this._entries.get(scope);
    if (existing !== undefined) {
      return existing;
    }
    const entries: string[] = [];
    this._entries.set(scope, entries);
    return entries;
  }
}

// ---------------------------------------------------------------------------
// Scoring helpers
// ---------------------------------------------------------------------------

/**
 * Score entries by word overlap with pre-computed query words.
 *
 * @param queryWords - Lowercase word set from the search query.
 * @param entries - Candidate knowledge strings.
 * @returns List of [score, entry] tuples above the relevance threshold.
 */
function scoreEntries(
  queryWords: ReadonlySet<string>,
  entries: readonly string[],
): Array<{ score: number; entry: string }> {
  const scored: Array<{ score: number; entry: string }> = [];
  for (const entry of entries) {
    const score = wordOverlapScore(queryWords, entry);
    if (score >= MIN_RELEVANCE_SCORE) {
      scored.push({ score, entry });
    }
  }
  return scored;
}

/**
 * Sort scored entries descending and return top-k texts.
 *
 * @param scored - List of scored entries.
 * @param maxResults - Maximum entries to return.
 * @returns Entry texts sorted by relevance, capped at maxResults.
 */
function selectTopResults(
  scored: Array<{ score: number; entry: string }>,
  maxResults: number,
): string[] {
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, maxResults).map((item) => item.entry);
}

/**
 * Compute word-overlap ratio between query words and text.
 *
 * @param queryWords - Pre-computed lowercase word set from the query.
 * @param text - Text to score against the query.
 * @returns Ratio of overlapping words to total query words (0.0 to 1.0).
 */
function wordOverlapScore(queryWords: ReadonlySet<string>, text: string): number {
  const textWords = toWordSet(text);
  if (textWords.size === 0) {
    return 0.0;
  }
  let overlap = 0;
  for (const word of queryWords) {
    if (textWords.has(word)) {
      overlap++;
    }
  }
  return overlap / queryWords.size;
}

/**
 * Split text into a lowercase word set.
 *
 * @param text - Input text.
 * @returns Set of lowercase words (whitespace-split).
 */
function toWordSet(text: string): Set<string> {
  const trimmed = text.trim();
  if (!trimmed) {
    return new Set();
  }
  return new Set(trimmed.toLowerCase().split(/\s+/));
}
